lib/sources.nix

```diff
@@ -26,12 +26,28 @@ rec {
     (type == "symlink" && lib.hasPrefix "result" baseName)
   );
 
-  cleanSource = builtins.filterSource cleanSourceFilter;
+  cleanSource = filterSource cleanSourceFilter;
+
+  # Like `builtins.filterSource`, except it will compose with itself,
+  # allowing you to chain multiple calls together without any
+  # intermediate copies being put in the nix store.
+  filterSource = f: src:
+    let isPath = builtins.typeOf src == "path";
+    in assert isLocalSource src; rec {
+      filter = if isPath then f else name: type: f name type && src.filter name type;
+      origSrc = if isPath then src else src.origSrc;
+      outPath = builtins.filterSource filter origSrc;
+      __is-lib-filterSource = true;
```

Comment:
There's nothing standardized here but IMO double underscore should be reserved by nix itself.

Reply:
Single underscore then?

Reply:
Yeah.